### DISCUSSION TITLE
Compiling Tilck with TCC & self-hosting?

### DISCUSSION THREAD

**Discussion by vvaltchev** [vvaltchev/tilck#93] [Q&A] [answered]
From a youtube comment:

> Vlad, you mention that you got the Tiny C Compiler  to run under TILCK, and that TCC can compile the Linux kernel. Can TCC compile the TILCK kernel, too? It would be neat if your system was self-hosting! And if it currently isn't, what would it take to make it so? Just curious.

**Comment by vvaltchev** [comment, marked as answer]
Unfortuantely, Tilck _cannot_ be built with TinyCC out of the box. In addition to the requirement for the GAS assembler, the project uses some new C11 features like atomics that are not supported by TinyCC, along with a bunch of GCC extensions. With enough effort, it would be possible to make Tilck's C files to build with TinyCC, but a full build certainly will need more than that: Gas, CMake, Bash and several other tools need to work too. Also, Tilck uses linker scripts and I'm not sure TinyCC's embedded linker can support that. It's true that TinyCC was able to build Linux 2.4, but the build probably used several other tools, not just the compiler alone.

I guess that the shortest path for self-hosting would be to make the GCC toolchain current used + cmake and bash to work on Tilck. To do that, Tilck will probably need more syscalls to support glibc (currently apps use libmusl) and will need to support dynamic linking as well. But, that's not part of project's goals at the moment. Tilck is a small project aiming at _embedded systems_ and the educational world. There is a lot to play with, in particular around filesystems, block devices, page cache etc. before getting to point of considering self-hosting.

**Reply by kassane** [reply]
Thinking about the complexity of GCC, is it possible to compile using LLVM?

Unlike Kernel Linux, it seems to me that there is still resistance in joining the toolchain as an alternative to GCC. And what about the Tick would there be a problem?

In the readme it informs about the possibility of musl support. Particularly thought of trying something with zig (LLVM ) toolchain (built on musl[static]) and the ease of cross compilation.

**Reply by vvaltchev** [reply]
In theory, user-space applications could certainly be compiled with any toolchain, as along as the syscalls performed by the libc implementation are supported by the Tilck kernel and static linking is used. In practice, glibc uses too often new/advanced syscalls that Tilck does not support yet. Using LLVM or GCC doesn't matter for userspace: only the syscalls used matter.

About the kernel itself: the C and C++ source files part of the kernel can indeed be compiled by `clang`. I also have `clang` targets in the my CI pipelines but, unfortunately, it's not possible at the moment to compile the assembly files with `llvm-as` because it lacks some of the features of `gnu-as` (gas). Therefore, currently, a pure LLVM toolchain cannot compile Tilck because of that.
